libs/unity/library/Runtime/Scripts/NativeRender/NativeVideoRenderer.cs

```diff
@@ -0,0 +1,219 @@
+// Licensed under the MIT License.
+
+using System;
+using UnityEngine;
+using Microsoft.MixedReality.WebRTC.UnityPlugin;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.WebRTC
+{
+	/// <summary>
+	/// This will render the video stream through native calls with DX11 or OpenGL, completely bypassing C# marshalling.
+	/// This provides a considerable performance improvement.
+	/// </summary>
+    [AddComponentMenu("MixedReality-WebRTC/Native Video Renderer")]
+    public class NativeVideoRenderer : MonoBehaviour
+    {
+	    [SerializeField]
+	    private RawImage _rawImage;
+
+        private RemoteVideoTrack _source;
+        private Material _videoMaterial;
+        private NativeRenderer _nativeRenderer;
+
+        private Texture2D _textureY;
+        private Texture2D _textureU;
+        private Texture2D _textureV;
+
+        private VideoFrameQueue<I420AVideoFrameStorage> _i420aFrameQueue;
+
+        private void Awake()
+        {
+	        NativeRenderingPluginUpdate.AddRef(this);
+        }
+
+        private void OnDestroy()
+        {
+	        NativeRenderingPluginUpdate.DecRef(this);
+
+	        // Teardown will also get called when the PeerConnection is destroyed.
+	        // So this is probably going to be called twice in many cases.
+	        TearDown();
+        }
+
+        private void Update()
+        {
+            if (_i420aFrameQueue != null)
+            {
+	            if (_i420aFrameQueue.TryDequeue(out I420AVideoFrameStorage frame))
+	            {
+		            StartNativeRendering((int)frame.Width, (int)frame.Height);
+	            }
+            }
+        }
+
+	    /// <summary>
+        /// Start rendering the passed source.
+        /// </summary>
+        /// <remarks>
+        /// Can be used to handle <see cref="VideoTrackSource.VideoStreamStarted"/> or <see cref="VideoReceiver.VideoStreamStarted"/>.
+        /// </remarks>
+        public void StartRendering(IVideoSource source)
+        {
+	        _source = source as RemoteVideoTrack;
+	        Debug.Assert(_source != null, "NativeVideoRender currently only supports RemoteVideoTack");
+
+	        switch (source.FrameEncoding)
+            {
+                case VideoEncoding.I420A:
+	                _i420aFrameQueue = new VideoFrameQueue<I420AVideoFrameStorage>(2);
+	                _source.I420AVideoFrameReady += I420AVideoFrameReady;
+                    break;
+                case VideoEncoding.Argb32:
+	                break;
+            }
+        }
+
+        private void I420AVideoFrameReady(I420AVideoFrame frame)
+        {
+	        // Clear the video display to not confuse the user who could otherwise
+	        // think that the video is still playing but is lagging/frozen.
+	        _i420aFrameQueue.Enqueue(frame);
+        }
+
+        private void StartNativeRendering(int width, int height)
+        {
+	        // Subscription is only used to ge the frame dimensions to generate the textures. So Unsubscribe once that is done.
```

Comment:
Note that _in theory_ WebRTC can rescale the stream and change the frame dimensions from one frame to another. In practice I haven't seen it happening in my tests. So it's fine if this supports constant dimensions for now, but we should take care to fail cleanly (or at the very least crash with a clear error in a clear callstack) if the dimensions happen to change.

Reply:
I've seen it happening on 2 occasions:
- Network conditions degradation/recovering (hard to repro)
- On Android, simply rotate the phone from portrait to landscape, and that flips width/height.

Reply:
Frame dimensions can also vary when changing source without reconnecting the Peer, e.g. when going from webcam to a shared desktop screen. We change resolutions on average at least once in every call, so constant dimensions wouldn't work for us.

Reply:
I made it output a warning if frame dimensions change after initialization. Some research will need to be done to figure out how to let this texture be resized from the C++ side.

Reply:
This is good enough for the PR to go in, I will open an issue to improve this once that's done.

Reply:
Logged as #631